### DISCUSSION TITLE
[Proposal] allow is operator compare generic type

### DISCUSSION THREAD

**Discussion by leandromoh** [dotnet/csharplang#3432] [Language Ideas]
as well we can do `x is string` to check if the instance `x` is of type `string`, would be amazing extend this behaviour to generic types, like `T is string`.

`T is SomeType` would be equivalent to `typeof(T).IsAssignableTo(SomeType)`

**Comment by jnm2** [comment]
@leandromoh Why would it be amazing to be able to do `T is string` rather than `typeof(T) == typeof(string)`?

**Comment by YairHalberstadt** [comment]
Note these are not equivalent. `T is SomeType` would be equivalent to `typeof(T).IsAssignableTo(SomeType)`.

**Comment by leandromoh** [comment]
@YairHalberstadt yes, you are right about what the proposal means. I will fix the summary

@jnm2 it would avoid boilerplate code (like use `typeof` twice).

**Comment by jnm2** [comment]
@YairHalberstadt @leandromoh IsAssignableTo is in API review, hasn't been shipped: https://github.com/dotnet/runtime/issues/30714

@leandromoh I was going to disagree that `typeof` twice is enough boilerplate to blink at, but now that I assume it's going to be updated to `IsAssignableFrom` which is an API that I can't stand, I actually think you have a point. However, I don't often see IsAssignableFrom in code. Do you?

**Comment by leandromoh** [comment]
> I actually think you have a point. However, I don't often see IsAssignableFrom in code. Do you?

@jnm2 Yes, when dealing with generics and metaprogramming (like a lot of libraries does) comparing types is a common task. so should be nice to simplify theses cases:

1) from `typeof(T).IsAssignableTo(SomeType)` to `T is SomeType`

2) from `typeof(T) == SomeType` to `T == SomeType`
